cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -0,0 +1,632 @@
+      case TimeUnit::MICRO:
+        return day_of_year_zoned_caster_template<std::chrono::microseconds>(timezone);
+      case TimeUnit::NANO:
+        return day_of_year_zoned_caster_template<std::chrono::nanoseconds>(timezone);
+    }
+  }
+  return day_of_year_caster_template<std::chrono::seconds>;
+}
+
+template <typename Duration>
+inline unsigned week_caster_template(const int64_t data) {
+  // Based on
+  // https://github.com/HowardHinnant/date/blob/6e921e1b1d21e84a5c82416ba7ecd98e33a436d0/include/date/iso_week.h#L1503
+  const auto dp = sys_days{floor<days>(Duration{data})};
+  auto y = year_month_day{dp + days{3}}.year();
+  auto start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+  if (dp < start) {
+    --y;
+    start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+  }
+  return static_cast<unsigned>(trunc<weeks>(dp - start).count() + 1);
+}
+
+template <typename Duration>
+inline std::function<unsigned(const int64_t)> week_zoned_caster_template(
+    const std::string timezone) {
+  static const arrow_vendored::date::time_zone* tz = locate_zone(timezone);
+  return [](const int64_t data) {
+    const auto ld = floor<days>(tz->to_local(sys_time<Duration>(Duration{data})));
+    auto y = year_month_day{ld + days{3}}.year();
+    auto start = local_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    if (ld < start) {
+      --y;
+      start = local_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    }
+    return static_cast<unsigned>(trunc<weeks>(local_days(ld) - start).count() + 1);
+  };
+}
+
+inline std::function<unsigned(const int64_t)> make_week_caster(
+    const std::shared_ptr<DataType> type) {
+  const auto ts_type = std::static_pointer_cast<const TimestampType>(type);
+  const TimeUnit::type unit = ts_type->unit();
+  const std::string timezone = ts_type->timezone();
+
+  if (timezone.empty()) {
+    switch (unit) {
+      case TimeUnit::SECOND:
+        return week_caster_template<std::chrono::seconds>;
+      case TimeUnit::MILLI:
+        return week_caster_template<std::chrono::milliseconds>;
+      case TimeUnit::MICRO:
+        return week_caster_template<std::chrono::microseconds>;
+      case TimeUnit::NANO:
+        return week_caster_template<std::chrono::nanoseconds>;
+    }
+  } else {
+    switch (unit) {
+      case TimeUnit::SECOND:
+        return week_zoned_caster_template<std::chrono::seconds>(timezone);
+      case TimeUnit::MILLI:
+        return week_zoned_caster_template<std::chrono::milliseconds>(timezone);
+      case TimeUnit::MICRO:
+        return week_zoned_caster_template<std::chrono::microseconds>(timezone);
+      case TimeUnit::NANO:
+        return week_zoned_caster_template<std::chrono::nanoseconds>(timezone);
+    }
+  }
+  return day_of_year_caster_template<std::chrono::seconds>;
+}
+
+// ----------------------------------------------------------------------
+// Extract year from timestamp
+
+template <typename out_type>
```

Comment:
As per the coding conventions, this should probably `OutType`.

Reply:
Done.